### PR TITLE
Feature/upload own data frontend

### PR DESCRIPTION
**Pull request by cephaschapa**
Changes
- Adds file input component that allows drag and drop
- Adds file card component
- Adds review page showing files added for each sector
- Adds redux-persist to manage file data this includes root state
- Adds translation to i18n
- Adds redux slice to manage inventory file data by sector
- modified redux store
- adds helper function to handle file data

### CODE REVIEW

File: app/src/app/[lng]/data/review/page.tsx [hunk 1]
```diff
@@ -0,0 +1,565 @@
...
+        formData.append("status", fileData.status!);
+        formData.append("fileReference", "");
+        formData.append("url", fileData.url!);
+        formData.append("gpcRefNo", "");
+        formData.append("data", file, file.name);
+      }
+
+      await addUserFile(formData).then(() => {
+        // TODO
+        // Trigger notification to user
```
Comment (lemilonkh) [inline app/src/app/[lng]/data/review/page.tsx:85]: Should we add a toast here?

File: app/src/app/[lng]/data/review/page.tsx [hunk 2]
```diff
@@ -0,0 +1,565 @@
...
+        dispatch(clear());
+        router.push("/");
```
Comment (lemilonkh) [inline app/src/app/[lng]/data/review/page.tsx:87]: This is weird to me. There can be multiple files, so either await Promise.all or do the router/ clear logic somewhere else.

File: app/src/app/[lng]/data/review/page.tsx [hunk 3]
```diff
@@ -0,0 +1,565 @@
...
+                  </Text>
+                </Box>
+              </Box>
+              <Box display="flex" gap="16px">
+                <Button
+                  color="sentiment.negativeDefault"
+                  variant="ghost"
+                  w="298px"
+                  gap="8px"
+                  onClick={() => onDiscardSectorChanges("Stationary Energy")}
```
Comment (lemilonkh) [inline app/src/app/[lng]/data/review/page.tsx:169]: I would prefer identifying sectors with their GPC reference number (I/ II/ III) or their ID

File: app/src/app/[lng]/data/review/page.tsx [hunk 4]
```diff
@@ -0,0 +1,565 @@
...
+              >
+                {t("third-party-data")} (0)
+              </Text>
+              <Box
+                display="grid"
+                gridTemplateColumns="auto auto auto"
+                gap="8px"
+              >
+                {/* TODO: show connected third party data */}
+                {/* <ThirdPartyDataCard /> */}
```
Comment (lemilonkh) [inline app/src/app/[lng]/data/review/page.tsx:221]: Should this be done here or in a future PR?

File: app/src/app/[lng]/data/review/page.tsx [hunk 5]
```diff
@@ -0,0 +1,565 @@
...
+                    {waterAndWasteWater[0].files.length})
+                  </Text>
+                  <Box
+                    display="grid"
+                    gridTemplateColumns="auto auto auto"
+                    gap="8px"
+                  >
+                    {waterAndWasteWater[0]?.files.map(
+                      (file: any, i: number) => <FileDataCard key={i} />,
+                    )}
```
Comment (lemilonkh) [inline app/src/app/[lng]/data/review/page.tsx:522]: You're not using the file here? Don't you need to show the file name etc.?

File: app/src/app/[lng]/data/review/page.tsx [hunk 6]
```diff
@@ -0,0 +1,565 @@
...
+                mr={4}
+                borderWidth="2px"
+                borderColor="sentiment.negativeDefault"
+                color="sentiment.negativeDefault"
+              >
+                {t("discard-all-changes")}
+              </Button>
+              <Button
+                h={16}
+                // isLoading={isConfirming}
```
Comment (lemilonkh) [inline app/src/app/[lng]/data/review/page.tsx:552]: Would be good to have this as user feedback here.

File: app/src/app/[lng]/data/review/page.tsx [hunk 7]
```diff
@@ -0,0 +1,565 @@
...
+                    {t("data-files-uploaded")}({transportation[0].files.length})
+                  </Text>
+                  <Box
+                    display="grid"
+                    gridTemplateColumns="auto auto auto"
+                    gap="8px"
+                  >
+                    {transportation[0]?.files.map((file: any, i: number) => (
+                      <FileDataCard key={i} />
+                    ))}
```
Comment (lemilonkh) [inline app/src/app/[lng]/data/review/page.tsx:382]: Use file to show its information?

File: app/src/app/[lng]/data/[step]/page.tsx [hunk 8]
```diff
@@ -399,7 +421,7 @@ export default function AddDataSteps({
     await new Promise((resolve) => setTimeout(resolve, 1000));
     setConfirming(false);
     if (activeStep >= steps.length - 1) {
-      router.push("/"); // go back to dashboard until there is a confirmation page
+      router.push("/data/review"); // go back to dashboard until there is a confirmation page
```
Comment (lemilonkh) [inline app/src/app/[lng]/data/[step]/page.tsx:424]: Comment can be removed now

File: app/src/app/[lng]/data/[step]/page.tsx [hunk 9]
```diff
@@ -408,7 +430,7 @@ export default function AddDataSteps({
 
   const onSkip = () => {
     if (activeStep >= steps.length - 1) {
-      router.push("/"); // go back to dashboard until there is a confirmation page
+      router.push("/data/review"); // go back to dashboard until there is a confirmation page
```
Comment (lemilonkh) [inline app/src/app/[lng]/data/[step]/page.tsx:433]: Comment can be removed now

File: app/src/app/[lng]/data/[step]/page.tsx [hunk 10]
```diff
@@ -417,6 +439,60 @@ export default function AddDataSteps({
...
+  const dispatch = useDispatch();
+  function fileToBase64(file: File) {
+    return new Promise((resolve, reject) => {
+      const reader = new FileReader();
+      reader.readAsDataURL(file);
+
+      reader.onload = () => resolve(reader.result);
+      reader.onerror = (error) => reject(error);
+    });
+  }
```
Comment (lemilonkh) [inline app/src/app/[lng]/data/[step]/page.tsx:457]: Not sure if transmitting files as base64 is the best. It works for now, but transmitting it as a binary buffer (which imo is valid for form data) is probably better, especially for larger files.

File: app/src/app/[lng]/data/[step]/page.tsx [hunk 11]
```diff
@@ -417,6 +439,60 @@ export default function AddDataSteps({
...
+      append({
+        sectorName: currentStep.title!,
+        fileData: {
+          uniqueFileId: uuidv4(),
+          userId: userInfo?.userId,
+          sector: currentStep.title,
+          data: base64FileString,
+          url: URL,
+          fileType: file.name.slice(
+            ((filename.lastIndexOf(".") - 1) >>> 0) + 2,
```
Comment (lemilonkh) [inline app/src/app/[lng]/data/[step]/page.tsx:472]: What is the right shift here? This seems a bit flaky.

File: app/src/components/Tabs/my-files-tab.tsx [hunk 12]
```diff
@@ -238,10 +232,28 @@ const MyFilesTab: FC<MyFilesTabProps> = ({
...
+  function filterDataByYear(
+    data: any,
+    selectedYear: number | null | undefined,
+  ) {
+    return data?.filter(
+      (item: any) => getYearFromDate(item?.created) === selectedYear,
+    );
+  }
+  const [isYearSelected, setIsYearSelected] = useState<boolean>(false);
+  const [selectedYear, setselectedYear] = useState<number | null>();
```
Comment (lemilonkh) [inline app/src/components/Tabs/my-files-tab.tsx:254]: Rename to setSelectedYear

File: app/src/components/Tabs/my-files-tab.tsx [hunk 13]
```diff
@@ -518,165 +529,155 @@ const MyFilesTab: FC<MyFilesTabProps> = ({
...
+                                              alignItems="center"
+                                              px="16px"
+                                              paddingTop="12px"
+                                              paddingBottom="12px"
+                                              _hover={{
+                                                background: "content.link",
+                                                color: "white",
+                                              }}
+                                              onClick={() => {
+                                                alert(city.id);
```
Comment (lemilonkh) [inline app/src/components/Tabs/my-files-tab.tsx:625]: Alert can maybe be removed?

File: app/src/lib/store.ts [hunk 14]
```diff
@@ -1,24 +1,54 @@
...
+  storage,
+};
+
+const reducer = combineReducers({
+  inventoryData: inventoryDataReducer,
+  [api.reducerPath]: api.reducer,
+  [openclimateAPI.reducerPath]: openclimateAPI.reducer,
+  city: cityReducer,
+  openClimateCity: openclimateCityReducer,
+  openClimateCityData: openclimateCityDataReducer,
```
[the remark this answers is not in the document]
Comment (cephaschapa) [reply app/src/lib/store.ts:33]: i separated the openclimate city data here to get the contextual data you need a different api call.

File: app/src/app/[lng]/data/review/page.tsx [hunk 15]
```diff
@@ -62,30 +55,43 @@ export default function ReviewPage({
...
+            fileData.data,
+            `${fileData.fileName}`,
+          );
+          formData.append("userId", fileData.userId!);
+          formData.append("fileName", fileData.fileName!);
+          formData.append("sector", fileData.sector!);
+          formData.append("status", defaultStatus);
+          formData.append("fileReference", "");
+          formData.append("url", fileData.url!);
+          formData.append("gpcRefNo", "");
```
Comment (lemilonkh) [inline app/src/app/[lng]/data/review/page.tsx:78]: This should be set to the reference number of the current sector? So either I, II or III.

File: app/src/components/Modals/delete-file-modal.tsx [hunk 16]
```diff
@@ -84,9 +94,8 @@ const DeleteFileModal: FC<DeleteFileModalProps> = ({
                 >
                   Are you sure you want to{" "}
                   <span style={{ fontWeight: "bold" }}>
-                    permanently remove {userData.name}
+                    permanently remove this file
```
Comment (lemilonkh) [inline app/src/components/Modals/delete-file-modal.tsx:97]: i18n for this modal? Can also be done in a follow-up PR.

File: app/tests/api/userfile.test.ts [hunk 17]
```diff
@@ -84,7 +84,7 @@ describe("UserFile API", () => {
     const res = await createUserFile(req, { params: { user: testUserID } });
     assert.equal(res.status, 200);
     const { data } = await res.json();
-
+    console.log(data);
```
Comment (lemilonkh) [inline app/tests/api/userfile.test.ts:87]: Remove logging

**Review by lemilonkh — APPROVED** [review]

**Pull request merged**